### PR TITLE
mds: add read/write io size metrics support

### PR DESCRIPTION
**Pull request by lxbsz**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by trociny — COMMENTED** [review]
Am I right understanding that all these changes are need for to be able to provide min_size and max_size counters?

Because for "total_ios" and "total_size" you could use standard "ops" and "bytes" counters. And actually I would prefer this way so we use the same convention everywhere.

As for max/min size, I wonder how useful it may be? histograms are useful,  but just min/max...

And average can be obtained from `bytes / ops`.

File: src/mgr/MDSPerfMetricCollector.cc [hunk 1]
```diff
@@ -26,8 +26,12 @@ void MDSPerfMetricCollector::process_reports(const MetricPayload &payload) {
   std::lock_guard locker(lock);
   process_reports_generic(
     metric_report.reports, [](PerformanceCounter *counter, const PerformanceCounter &update) {
-      counter->first = update.first;
-      counter->second = update.second;
+      if (counter->size() < update.size()) {
+        counter->resize(2);
```
[the remark this answers is not in the document]
Comment (lxbsz) [reply src/mgr/MDSPerfMetricCollector.cc]: Yeah, I have fixed it in the last patch, I will fix it here.

**Comment by lxbsz** [comment]
There still need some work on the `cephfs-top` tool to correctly show the size metrics.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by trociny** [comment]
> Just switched the pair to vector to make sure we can transfer more than 2 numbers for each metric type.

I am fine with making it vector if we need to extend it. I just still don't see a need to do this for this particular case, because I am not convinced you need the counters for min/max. It does not look very useful for me. It should be a histogram if you want it to be really useful.

> Currently it will show min/max/total/ops, maybe we should switch to min/max/avg/ops instead.

Usually we collect total and ops, and avg is calculated as total / ops if needed.

**Comment by lxbsz** [comment]
> > Just switched the pair to vector to make sure we can transfer more than 2 numbers for each metric type.
> 
> I am fine with making it vector if we need to extend it. I just still don't see a need to do this for this particular case, because I am not convinced you need the counters for min/max. It does not look very useful for me. It should be a histogram if you want it to be really useful.

Yeah, I have remove it for now and currently we will report the total size and total ops.
I kept the switching the pair to vector patch in this PR and we need it when we add the io speed later soon in this PR later or a next PR.

> 
> > Currently it will show min/max/total/ops, maybe we should switch to min/max/avg/ops instead.
> 
> Usually we collect total and ops, and avg is calculated as total / ops if needed.

Yeah, sounds reasonable.

**Comment by lxbsz** [comment]
The output will be:
```
cephfs-top - Wed Mar 31 22:09:19 2021                                                                                                     
Client(s): 1 - 1 FUSE, 0 kclient, 0 libcephfs                                                                                             
                                                                                                                                          
  client_id mount_root chit(%) rlat(s) wlat(s) mlat(s) dlease(%) ofiles oicaps oinodes rio(Kb) wio(Kb) mount_piont@host/addr              
  4637      /          100.0   12.18   25.61   0.06    0.0       0      2      0       3.69    126.33  /mnt/cephfs@lxbceph1/v1:10.72.47.117                                                                                                                                        
                    
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test api

**Comment by lxbsz** [comment]
> > I kept the switching the pair to vector patch in this PR and we need it when we add the io speed later soon in this PR later or a next PR.
> 
> Why do you need special counters for io speed? Usually you just collect total_bytes, total_ops and then the speed can be calculated as a difference between two collections divided by the collection time period.

Yeah, this is what I did locally, but I am doing it in the MDS daemon collecter instead, not the `cephfs-top` tool. Because for the manually query command `./bin/ceph fs perf stats`, we cannot hold the previous collection.

The cephfs userspace clients and kernel clients will collect and report the metrics to MDS daemon per second as usually, but could change. So the client will report ops/total_size and the duration from the last time to report metrics to MDS daemon collecter, and the MDS daemon collecter will compute the speeds and avg_size and then report ops/avg_size/speed to cephfs-top tool.

But if we do not need to show the speeds in the output of `./bin/ceph fs perf stats` command mentioned above, then the https://github.com/ceph/ceph/pull/40514/commits/7912e746b9d91d22fb307cc4149fbe683e4f9f35 commit could be removed. And then we will only compute and show the speeds in the `cephfs-top` tool.

@vshankar Any idea or advice about this ?

**Comment by lxbsz** [comment]
This depends on https://github.com/ceph/ceph/pull/40539, which is fixing the `cephfs-top` tool.

**Comment by vshankar** [comment]
(sorry for the latency in replying)

> > > I kept the switching the pair to vector patch in this PR and we need it when we add the io speed later soon in this PR later or a next PR.
> > 
> > 
> > Why do you need special counters for io speed? Usually you just collect total_bytes, total_ops and then the speed can be calculated as a difference between two collections divided by the collection time period.
> 
> Yeah, this is what I did locally, but I am doing it in the MDS daemon collecter instead, not the `cephfs-top` tool. Because for the manually query command `./bin/ceph fs perf stats`, we cannot hold the previous collection.
> 
> The cephfs userspace clients and kernel clients will collect and report the metrics to MDS daemon per second as usually, but could change. So the client will report ops/total_size and the duration from the last time to report metrics to MDS daemon collecter, and the MDS daemon collecter will compute the speeds and avg_size and then report ops/avg_size/speed to cephfs-top tool.

If I'm understanding this correctly, wouldn't it suffice for the report this counter calculated at fixed interval irrespective of its metrics reporting tick?

> 
> But if we do not need to show the speeds in the output of `./bin/ceph fs perf stats` command mentioned above, then the [7912e74](https://github.com/ceph/ceph/commit/7912e746b9d91d22fb307cc4149fbe683e4f9f35) commit could be removed. And then we will only compute and show the speeds in the `cephfs-top` tool.
> 
> @vshankar Any idea or advice about this ?

**Comment by lxbsz** [comment]
> If I'm understanding this correctly, wouldn't it suffice for the report this counter calculated at fixed interval irrespective of its metrics reporting tick?
> 

Do you mean in the `Client::tick()`, right ?

If so, I am afraid it won't work for the kclient, because in kernel space it's not easy and not recommanded to compute with float type.

**Comment by vshankar** [comment]
> Do you mean in the `Client::tick()`, right ?
> 
> If so, I am afraid it won't work for the kclient, because in kernel space it's not easy and not recommanded to compute with float type.

Ah ok! That's right. So, the option left is to calculate it in the MDS or cephfs-top tool. There is no hard reason to show this in `perf stats` (there are no other direct consumers of this right now other than cephfs-top).

**Comment by lxbsz** [comment]
> > Do you mean in the `Client::tick()`, right ?
> > If so, I am afraid it won't work for the kclient, because in kernel space it's not easy and not recommanded to compute with float type.
> 
> Ah ok! That's right. So, the option left is to calculate it in the MDS or cephfs-top tool. There is no hard reason to show this in `perf stats` (there are no other direct consumers of this right now other than cephfs-top).

Okay, that's cool. So let's compute it in the cephfs-top tool.

**Comment by lxbsz** [comment]
Updated it:

- calculate the read/write IO speeds in cephfs-top instead.
- remove the 7912e74 commit.

File: doc/man/8/cephfs-top.rst [hunk 2]
```diff
@@ -71,6 +71,22 @@ Descriptions of fields
 
    number of opened inodes
 
+.. describe:: rio
+
+   average size of read IOs
+
+.. describe:: wio
+
+   average size of write IOs
```
Comment (batrick) [inline doc/man/8/cephfs-top.rst:80]: How about also adding a cumulative bytes read/written?

File: src/tools/cephfs/top/cephfs-top [hunk 3]
```diff
@@ -206,6 +229,24 @@ class FSTop(object):
             return "(%)"
         elif typ == MetricType.METRIC_TYPE_LATENCY:
             return "(s)"
+        elif typ == MetricType.METRIC_TYPE_SIZE:
+            return "(Kb)"
```
Comment (batrick) [inline src/tools/cephfs/top/cephfs-top:233]: `(kb)` is probably too small; in a large cluster it will look silly. How about MB/s?

**Review by batrick — CHANGES_REQUESTED** [review]
otherwise lgtm

File: src/tools/cephfs/top/cephfs-top (hunk 3 above)
Comment (lxbsz) [reply src/tools/cephfs/top/cephfs-top:233]: LGTM, will fix it.

File: doc/man/8/cephfs-top.rst (hunk 2 above)
Comment (lxbsz) [reply doc/man/8/cephfs-top.rst:80]: Make sense, how about:

```
... rio(Mb)   ...   wio(Mb)
... sum/avg ...   sum/avg
```

File: doc/man/8/cephfs-top.rst [hunk 4]
```diff
@@ -71,6 +71,30 @@ Descriptions of fields
 
    number of opened inodes
 
+.. describe:: rio
+
+   total size of read IOs
+
+.. describe:: wio
```
Comment (batrick) [inline doc/man/8/cephfs-top.rst:78]: `wtot` (write total) and `rtot` (read total) would be more intuitive I think. @vshankar @joscollin what do you think?

File: doc/man/8/cephfs-top.rst (hunk 2 above)
Comment (batrick) [reply doc/man/8/cephfs-top.rst:88]: https://github.com/ceph/ceph/pull/40514/files#r625471977

File: doc/man/8/cephfs-top.rst (hunk 4 above)
Comment (joscollin) [reply doc/man/8/cephfs-top.rst:78]: Looks okay, trivial to me. What about `rtio` and `wtio`?

Comment (lxbsz) [reply doc/man/8/cephfs-top.rst:78]: Yeah, the `rtio` and `wtio` seems better for me.

File: src/tools/cephfs/top/cephfs-top [hunk 5]
```diff
@@ -206,6 +237,33 @@ class FSTop(object):
...
+            return "rsp"
+        if item == "WRITE_IO_SIZES":
+            return "wsp"
+        else:
+            # return empty string for none type
+            return ''
+
+    def speed_mtype(self, typ):
+        if typ == MetricType.METRIC_TYPE_SIZE:
+            return "(Mb/s)"
```
Comment (batrick) [inline src/tools/cephfs/top/cephfs-top:266]:
```suggestion
            return "(MB/s)"
```

File: src/tools/cephfs/top/cephfs-top [hunk 6]
```diff
@@ -206,6 +237,33 @@ class FSTop(object):
             return "(%)"
         elif typ == MetricType.METRIC_TYPE_LATENCY:
             return "(s)"
+        elif typ == MetricType.METRIC_TYPE_SIZE:
+            return "(Mb)"
```
Comment (batrick) [inline src/tools/cephfs/top/cephfs-top:241]:
```suggestion
            return "(MB)"
```

Comment (batrick) [reply src/tools/cephfs/top/cephfs-top:241]: `Mb` is mega**bits** while `Mb` is mega**bytes**

File: src/tools/cephfs/top/cephfs-top (hunk 5 above)
Comment (lxbsz) [reply src/tools/cephfs/top/cephfs-top]: Done.

File: src/mds/MetricsHandler.cc [hunk 7]
```diff
@@ -349,7 +385,7 @@ void MetricsHandler::update_rank0() {
   for (auto p = client_metrics_map.begin(); p != client_metrics_map.end();) {
     // copy metrics and update local metrics map as required
     auto &metrics = p->second.second;
-    update_client_metrics_map.emplace(p->first, metrics);
+    update_client_metrics_map.emplace(p->first, std::move(metrics));
     if (metrics.update_type == UPDATE_TYPE_REFRESH) {
```
[the remark this answers is not in the document]
Comment (lxbsz) [reply src/mds/MetricsHandler.cc]: Removed this commit for now, previously there had no code will access it, and after syncing to upstream I missed it.

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-October-26

**Pull request merged**